### DISCUSSION TITLE
version 4.0.0-beta.4

### DISCUSSION THREAD

**Discussion by hvianna** [hvianna/audioMotion-analyzer#42] [Announcements]
### 🚨 v4 BREAKING CHANGES:

+ Removed deprecated properties:
  + ~`energy`~ - use [`getEnergy()`](README.md#getenergy-preset-startfreq-endfreq-) instead
  + ~`isLedDisplay`~ - use [`isLedBars`](README.md#isledbars-boolean-read-only) instead
  + ~`peakEnergy`~ - use [`getEnergy('peak')`](README.md#getenergy-preset-startfreq-endfreq-) instead
  + ~`showLeds`~ - use [`ledBars`](README.md#ledbars-boolean) instead

+ [`getBars()`](README.md#getbars) elements now contain an extra `freq` property (potentially breaking in TypeScript projects), which represents the bar's center frequency.
The range of each bar has also changed from version 3.6.0, with frequencies that were previously at the lower edge (`freqLo`) now placed at the center.

### Added:

in 4.0.0-beta.4:

+ [`channelLayout`](README.md#channellayout-string) property - adds option for dual channel combined spectrum, with both channel graphs overlaid;
+ [`gradientLeft`](README.md#gradientleft-string) and [`gradientRight`](README.md#gradientright-string) properties - allow to select different gradients for each channel;
+ `stereo` property has been DEPRECATED - use `channelLayout` instead.

in 4.0.0-beta.3:

+ [`frequencyScale`](README.md#frequencyscale-string) property - adds linear and perceptual (Bark/Mel) frequency scales (thanks [**@TF3RDL**](https://github.com/hvianna/audioMotion-analyzer/issues/30));
+ [`isBandsMode`](README.md#isbandsmode-boolean-read-only) read-only property.

in 4.0.0-beta.1:

+ [`linearBoost`](README.md#linearboost-number) property - amplify low energy values when using [linear amplitude](README.md#linearamplitude-boolean);
+ [`weightingFilter`](README.md#weightingfilter-string) property - select from five different weighting filters for improved spectrum visualization.

in 4.0.0-beta.0:

+ [`ansiBands`](README.md#ansibands-boolean) property - use ANSI/IEC preferred frequencies for the octave bands (thanks [**@jonathan-annett**](https://github.com/hvianna/audioMotion.js/issues/28));
+ [`linearAmplitude`](README.md#linearamplitude-boolean) property - use linear values instead of dB for spectrum amplitudes (thanks [**@TF3RDL**](https://github.com/hvianna/audioMotion-analyzer/issues/30));
+ [`noteLabels`](README.md#notelabels-boolean) property - display musical note labels instead of frequency values on the X axis.

### Fixed:

+ LED peaks showing below zero level when `reflexRatio` > 0.

### Changed and improved:

+ Custom gradients can now register a single color;
+ [`splitGradient`](README.md#splitgradient-boolean) now works for horizontal gradients in [`radial`](README.md#radial-boolean) visualization;
+ Optimized generation of octave bands, with more accurate bands range and center frequencies;
+ Frequency labels on the X axis now show different values depending on whether the analyzer is set to use the equal-tempered scale
(default) or the ANSI standard bands (`ansiBands == true`), to properly match the octave bands center frequencies;
+ [`maxFreq`](README.md#maxfreq-number) is now capped to half the AudioContext's sample rate (Nyquist frequency);
+ Code optimizations, minor fixes and clean-up;
+ Updated demos.

<hr /><em>This discussion was created from the release <a href='https://github.com/hvianna/audioMotion-analyzer/releases/tag/4.0.0-beta.4'>version 4.0.0-beta.4</a>.</em>

**Comment by hvianna** [comment]
The [demo pen](https://codepen.io/hvianna/pen/MWBgjMb) has been updated to show the new dual channel combined spectrum by default:

![image](https://user-images.githubusercontent.com/1033735/218330992-e961839c-56a2-4b4c-9b72-4ed8ca8e6587.png)

**Reply by TF3RDL** [reply]
Yeah, but currently it works best with semi-transparent graphs on the default `source-over` of `CanvasRenderingContext2D.globalCompositeOperation` because otherwise you can't see the left channel beside the peaks if the right channel is higher than left channel (especially the fully opaque bars)

BTW, it is a nice addition considering the combined channel spectrum visualization looks more like MMultiAnalyzer or other multi-channel spectrum analyzer VSTs and I think adding the ability to change `globalCompositeOperation` to something like `lighter` allows it to properly visualize this type of visualization with fully-opaque LED bars

**Reply by hvianna** [reply]
I'll have a look into adding this. Thanks for the suggestion!
